Task/weak numba dependency

johannvk opened the pull request:
This PR adds numba as a weak / optional dependency of `skchange`, instead of the hard dependency it is now.
To achieve this, we use  the `sktime` function `_check_soft_dependencies`, and return a passthrough version of the `@jit` and `@njit` decorators if `numba` is not installed.

Additionally, this PR enables the configuring of some `numba` default `@njit` parameters through the setting of environment variables.
Currently the supported parameters are `cache`, `parallel`, and `fastmath`, set by the corresponding environment variables `NUMBA_CACHE`, `NUMBA_PARALLEL`, and `NUMBA_FASTMATH`. 
The current implementation only sets the defaults, so if a user of the `numba` decorator specifies e.g. `fastmath=False`, their choice will not be overridden.

Tveten commented on pyproject.toml:
```diff
@@ -37,14 +37,18 @@ requires-python = ">=3.9,<3.13"
 
 [project.optional-dependencies]
+# numba - numba as an optional dependency, install via "pip install skchange[numba]"
+numba = [
```
Perfect! Could you also add numba to the all_extras dependency set?

Tveten commented on skchange/utils/numba/njit.py:
This looks great! One feature I would really like in addition, which I think is missing now, is the option to turn numba off entirely through environment variables. Or is it supported somehow already?

johannvk replied:
Yes it's possible to set the environment variable " NUMBA_DISABLE_JIT" (as provided by [numba](https://numba.pydata.org/numba-doc/dev/reference/envvars.html)) to disable JIT entirely, and then the decorator from numba will in effect call the python function instead of a jitted version.

Tveten commented on skchange/utils/numba/njit.py:
```diff
@@ -0,0 +1,151 @@
+"""Dispatch njit decorator used to isolate numba.
+
+Copied from sktime.utils.numba.njit
```
Could you add a description to the header docstring about how to use the environment variables to customise numba? That you need to make a .env file and what options are supported in what format

johannvk replied:
Will do :)

johannvk commented:
> Another thing: I see the patch coverage is pretty low. Do you have time to write a few tests for the new njit module @johannvk? They can be placed in skchange/utils/numba/tests

It took some trial and error, but I added tests for all the new configuration fuctionality now, and also added another test step to the Github Actions workflow to run all our tests with and without Numba jit-compilation.

Tveten commented on skchange/anomaly_detectors/tests/test_anomaly_detectors.py:
```diff
@@ -48,8 +53,34 @@ def test_collective_anomaly_detector_transform(Estimator):
         assert (labels.iloc[start : end + 1] == i + 1).all()
 
 
+def test_custom_anomaly_detector_transform():
```
What is the role of this test in addition to test_collective_anomaly_detector_transform?

johannvk replied:
It was needed to test a few lines in the ''anomaly_score_factory'', to handle the case where we do not pass in a string to specify the score we want to use, but send in a tuple with two functions. 

Those specific lines were no longer covered with the changes this PR makes.
